windows/src/engine/keyman32/kmprocessactions.cpp

```diff
@@ -0,0 +1,137 @@
+    keyboardOpts[0].value = cp;
+    keyboardOpts[1] = KM_KBP_OPTIONS_END;
+    km_kbp_status eventStatus = km_kbp_state_options_update(keyboardState, keyboardOpts);
+    if (eventStatus != KM_KBP_STATUS_OK)
+    {
+      // log warning "problem saving option for km_kbp_keyboard");
+      SendDebugMessageFormat(0, sdmGlobal, 0, "ProcessHook: Error %d saving option for keyboard [%s].", eventStatus, activeKeyboard->Name);
+    }
+
+    for (int i = 0; i < 2; i++) {
+      delete[] keyboardOpts[i].key;
+      delete[] keyboardOpts[i].value;
+    }
+    // Put the keyboard option into Windows Registry
+    if (actionItem->option != NULL && actionItem->option->key != NULL &&
+      actionItem->option->value != NULL)
+    {
+      // log"Saving keyboard option to registry");
+      SendDebugMessageFormat(0, sdmGlobal, 0, "ProcessHook: Saving option to registry for keybooard [%s].", activeKeyboard->Name);
+      LPWSTR value = new WCHAR[sizeof(actionItem->option->value) + 1];
+      wcscpy_s(value, sizeof(actionItem->option->value) + 1, reinterpret_cast<LPCWSTR>(actionItem->option->value));
+      SaveKeyboardOptionREGCore(activeKeyboard, reinterpret_cast<LPCWSTR>(actionItem->option->key), value);
+    }
+  }
+  return TRUE;
+}
+
+static BOOL processInvalidateContext(
+  AITIP* app,
+  km_kbp_state* keyboardState
+) {
+  km_kbp_context_clear(km_kbp_state_context(keyboardState));
+  app->ResetContext();
+  return TRUE;
+}
+
+ // TODO: 5011  make this function a new file
+BOOL ProcessActions(BOOL* emitKeyStroke)
+{
+  PKEYMAN64THREADDATA _td = ThreadGlobals();
+  if (!_td) return FALSE;
+
+  // Process the action items from the core. This actions will modify the windows context (AppContext).
+  // Therefore it is not required to copy the context from the core to the windows context.
+
+  for (auto act = km_kbp_state_action_items(_td->lpActiveKeyboard->lpActiveKBState, nullptr); act->type != KM_KBP_IT_END; act++) {
+    BOOL continueProcessingActions = TRUE;
+    switch (act->type)
+    {
+    case KM_KBP_IT_CHAR:
+      continueProcessingActions = processUnicodeChar(_td->app, act);
+      break;
+    case KM_KBP_IT_MARKER:
+      continueProcessingActions = processMarker(_td->app, act);
+      break;
+    case KM_KBP_IT_ALERT:
+      continueProcessingActions = processAlert(_td->app);
+      break;
+    case KM_KBP_IT_BACK:
+      continueProcessingActions = processBack(_td->app, act);
+      break;
+    case KM_KBP_IT_PERSIST_OPT:
+      continueProcessingActions = processPersistOpt(act, _td->lpActiveKeyboard->lpActiveKBState, _td->lpActiveKeyboard);
+      break;
+    case KM_KBP_IT_EMIT_KEYSTROKE:
+      *emitKeyStroke = TRUE;
+      continueProcessingActions = TRUE;
+      break;
+    case KM_KBP_IT_INVALIDATE_CONTEXT:
+      continueProcessingActions = processInvalidateContext(_td->app, _td->lpActiveKeyboard->lpActiveKBState);
+      break;
+    case KM_KBP_IT_END:
+      assert(false);
+      break;
```

Comment:
This is covered by the default case, and we should never come here because we break the loop if we encounter `KM_KBP_IT_END`, so I'd omit these three lines.

Reply:
How about this 
``    case KM_KBP_IT_END:
    // fallthrough
    default:
      assert(false); // NOT SUPPORTED
      break; ``
 It still says we what every case fo the enum handled, but also makes it clear to future developers we have considered KM_KBP_IT_END and deliberately are falling through

Reply:
(I edited your comment to show how to fixup the indents 🤣 )